kustomize/commands/edit/set/setbuildmetadata.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2022 The Kubernetes Authors.
+// SPDX-License-Identifier: Apache-2.0
+
+package set
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"sigs.k8s.io/kustomize/api/types"
+	"sigs.k8s.io/kustomize/kustomize/v4/commands/internal/kustfile"
+	"sigs.k8s.io/kustomize/kyaml/filesys"
+)
+
+type setBuildMetadataOptions struct {
+	buildMetadataOptions []string
+}
+
+// newCmdSetBuildMetadata sets options in the kustomization's buildMetada field.
+func newCmdSetBuildMetadata(fSys filesys.FileSystem) *cobra.Command {
+	var o setBuildMetadataOptions
+
+	cmd := &cobra.Command{
+		Use:   "buildmetadata",
+		Short: "Sets one or more buildMetadata options to the kustomization.yaml in the current directory",
+		Long: `Sets one or more buildMetadata options to the kustomization.yaml in the current directory.
+Existing options in the buildMetadata field will be replaced entirely by the new options set by this command.
+The following options are valid:
+  - originAnnotations
+  - transformerAnnotations
+  - managedByLabel
+originAnnotations will add the annotation config.kubernetes.io/origin to each resource, describing where 
+each resource originated from.
+transformerAnnotations will add the annotation alpha.config.kubernetes.io/transformations to each resource,
+describing the transformers that have acted upon the resource.
+managedByLabel will add the label app.kubernetes.io/managed-by to each resource, describing which version
+of kustomize managed the resource.`,
+		Example: `
+		set buildmetadata {option1},{option2}`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			err := o.Validate(args)
+			if err != nil {
+				return err
+			}
+			return o.RunSetBuildMetadata(fSys)
+		},
+	}
+	return cmd
+}
+
+// Validate validates setBuildMetadata command.
+func (o *setBuildMetadataOptions) Validate(args []string) error {
+	if len(args) == 0 {
+		return errors.New("must specify a buildMetadata option")
+	}
+	if len(args) > 1 {
+		return fmt.Errorf("too many arguments: %s; to provide multiple buildMetadata options, please separate options by comma", args)
+	}
+	o.buildMetadataOptions = strings.Split(args[0], ",")
+	for _, opt := range o.buildMetadataOptions {
+		if !kustfile.StringInSlice(opt, types.BuildMetadataOptions) {
+			return fmt.Errorf("invalid buildMetadata option: %s", opt)
+		}
+	}
+	return nil
+}
```

Comment:
Normally we define the function `Validate` to be sharable among `setBuildMetadataOptions`,  `addBuildMetadataOptions`, `removeBuildMetadataOptions`? But I can see it introduces more code structure to `kustomize/commands/set` dir.

Do you think we can add a new structure buildMetadataValidator which has the Valdiate func, and itself is a attribtue to the other functions? e.g.

```go
package cmdvalidate // a sub dir under kustomize/commands/internal
type buildMetadataValidator struct {}

func (b *buildMetadataValidator) Validate() {
   ...
}
```
```go 

type setBuildMetadataOptions struct {
       *buildMetadataValidator
	buildMetadataOptions []string
}
```

Reply:
I did this in `kustomize/commands/util/validate.go`, because the util package already had some common functions shared among the edit commands. Thanks for the suggestion!